### PR TITLE
health_check: add Cached custom health checker

### PR DESCRIPTION
**Pull request by hudayou**
Signed-off-by: Jacky Hu <jacky.hu@walmart.com>

Commit Message: health_check: add Cached custom health checker
Additional Description: Allow use of cached  health check result from an external third party health checker
Risk Level: L
Testing: unit test, manual testing
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by hudayou** [comment]
Need to fix some format errors
/wait

**Comment by hudayou** [comment]
Need to figure out:
1. how to fix windows build
2. why compile_time_options ci failed?
3. add more tests for code coverage

/wait

File: api/envoy/extensions/health_checkers/cached/v3/cached.proto [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/health_checkers/cached/v3;cachedv3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Cached]
+// Cached health checker :ref:`configuration overview <config_health_checkers_cached>`.
+// [#extension: envoy.health_checkers.cached]
+
+// [#next-free-field: 9]
+message Cached {
```
Comment (adisuissa) [inline api/envoy/extensions/health_checkers/cached/v3/cached.proto:21]: Can/should the cached server be an upstream cluster?
This should take care of all the connection parameters.

**Review by adisuissa — COMMENTED** [review]
Thanks.
Can you provide some more background about this extension in the PR description (also is there a tracking issue)?

**Comment by hudayou** [comment]
> Thanks. Can you provide some more background about this extension in the PR description (also is there a tracking issue)?

Yep, we are using a redis cache to avoid excessive/unnecessary health checks to our backend servers.

Below is some description about it.

I'll update the PR description about it. There is no tracking issue, if needed by you I can create one.

https://github.com/envoyproxy/envoy/blob/badad5f2aadde7523be4e422146ff10a099055db/docs/root/configuration/upstream/health_checkers/cached.rst

File: api/envoy/extensions/health_checkers/cached/v3/cached.proto (hunk 1 above)
Comment (hudayou) [reply api/envoy/extensions/health_checkers/cached/v3/cached.proto:21]: No, this parameters are dedicated to a redis cache server
Which is handled by the implementation of this PR.

https://github.com/envoyproxy/envoy/blob/badad5f2aadde7523be4e422146ff10a099055db/source/extensions/health_checkers/cached/hiredis.cc

**Comment by hudayou** [comment]
Need to fix windows build issue by providing alias to static libs  as required
/wait

**Comment by mattklein123** [comment]
Hi,

Thanks for the contribution. We can't accept a contribution this large without a tracking issue, small design doc, etc. Also, this extension would need a sponsor to go in the main repo vs. contrib. Please read the contributing guide. I'm going to close this for now until we sort that out.

Thanks

**Comment by hudayou** [comment]
Sorry, I apologize for not providing a issue, I'll provide it today.

**Comment by hudayou** [comment]
Created issue as required here.

https://github.com/envoyproxy/envoy/issues/24757

**Pull request closed without merging**